### PR TITLE
executor: add XAPI executor.

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @zimulala PTAL

**Comment by shenli** [comment]
@nieyy @liugangnhm

File: executor/executor_x.go [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+			var err error
+			e.subResult, err = e.result.Next()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if e.subResult == nil {
+				return nil, nil
+			}
+		}
+		h, rowData, err := e.subResult.Next()
```
Comment (shenli) [inline executor/executor_x.go]: We may support join in the future. In that case, there will be multiple handles for a rowData.

**Comment by nieyy** [comment]
LGTM, but could you plz give us an intuitive example to demonstrate how the new method `PartialNext` used for?

File: executor/executor_x.go (hunk 1 above)
Comment (coocood) [inline executor/executor_x.go]: I think join is too complex for region server, and the benefit is not clear.

**Comment by coocood** [comment]
@nieyy `indexRangesToPBRangesAndPoints` uses `PartialNext`to convert a range with low/high exclude information to a range with closed low and open high range.

File: executor/executor_x.go [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+func resultFieldsToPBExpression(fields []*ast.ResultField) []*tipb.Expr {
+	return nil
+}
+
+func tableRangeToPBRangesAndPoints(tableRanges []plan.TableRange) ([]*tipb.KeyRange, [][]byte) {
+	hrs := make([]*tipb.KeyRange, 0, len(tableRanges))
+	var handlePoints [][]byte
+	for _, ran := range tableRanges {
+		if ran.LowVal == ran.HighVal {
```
Comment (shenli) [inline executor/executor_x.go]: Add IsPoint() function for TableRange? Just like IndexRange.

Comment (coocood) [inline executor/executor_x.go]: This code is too simple for create a function.

File: util/codec/codec.go [hunk 3]
```diff
@@ -22,14 +22,15 @@ import (
...
 const (
-	nilFlag byte = iota
-	bytesFlag
-	compactBytesFlag
-	intFlag
-	uintFlag
-	floatFlag
-	decimalFlag
-	durationFlag
+	nilFlag          byte = 0
```
Comment (zimulala) [inline util/codec/codec.go]: Here why not use `iota` ?

File: executor/executor_x.go [hunk 4]
```diff
@@ -0,0 +1,341 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (shenli) [inline executor/executor_x.go]: Is there any better name than executor_x.go? Such as executor_distribute.go.

File: executor/executor_x.go [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+		indexOrder[h] = i
+	}
+	sort.Sort(int64Slice(handles))
+	tblResult, err := e.doTableRequest(txn, handles)
+	unorderedRows, err := extractRowsFromTableResult(e.table, tblResult)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// Restore the original index order.
+	rows := make([]*Row, len(handles))
```
Comment (shenli) [inline executor/executor_x.go]: If the SQL statement do not specify result order, this operation is not necessary.

File: executor/executor_x.go (hunk 4 above)
Comment (coocood) [inline executor/executor_x.go]: That's too long, how about executor_xapi.go?

File: util/codec/codec.go (hunk 3 above)
Comment (coocood) [inline util/codec/codec.go]: To make the codec flag stable, we may need to add more flag later, insert in the middle will change many existing flag values.

File: executor/executor_x.go (hunk 4 above)
Comment (shenli) [inline executor/executor_x.go]: It is better.

File: executor/executor_x.go (hunk 5 above)
Comment (coocood) [inline executor/executor_x.go]: index scan without order assumption is exceptional, this should be the default behavior.
We can add an option for out of order index scan later, but that require changes on optimizer.

File: executor/executor_x.go (hunk 1 above)
Comment (nieyy) [inline executor/executor_x.go]: As I know, MySQL Cluster has pushed partial `Join` operations down, and gained some benefits.

File: executor/executor_x.go (hunk 4 above)
Comment (nieyy) [inline executor/executor_x.go]: I prefer `executor_xapi.go` also, :)

File: executor/executor_x.go [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+func conditionsToPBExpression(expr ...ast.ExprNode) *tipb.Expr {
+	return nil
+}
+
+func resultFieldsToPBExpression(fields []*ast.ResultField) []*tipb.Expr {
+	return nil
+}
+
+func tableRangeToPBRangesAndPoints(tableRanges []plan.TableRange) ([]*tipb.KeyRange, [][]byte) {
```
Comment (zimulala) [inline executor/executor_x.go]: tableRanges -> ranges like `indexRangesToPBRangesAndPoints` function?

File: executor/builder.go [hunk 7]
```diff
@@ -117,7 +120,33 @@ func (b *executorBuilder) buildFilter(src Executor, conditions []ast.ExprNode) E
 }
 
 func (b *executorBuilder) buildTableScan(v *plan.TableScan) Executor {
+	txn, err := b.ctx.GetTxn(false)
+	if err != nil {
+		b.err = err
+		return nil
+	}
 	table, _ := b.is.TableByID(v.Table.ID)
+	client := txn.GetClient()
```
Comment (liugangnhm) [inline executor/builder.go]: should we check  nil client?

Comment (coocood) [inline executor/builder.go]: No, every implementation should return a valid Client.

**Comment by zimulala** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

File: executor/builder.go (hunk 7 above)
Comment (liugangnhm) [inline executor/builder.go]: maybe we should add an panic:

**Pull request merged**